src/inference/graph_inference.cpp

```diff
@@ -786,6 +879,137 @@ class GraphNodeAssignment : public Nice2Assignment {
     }
   }
 
+  // Perform optimization based on factor features
+  void LocalPerFactorOptimizationPass(const GraphInference& fweights, size_t beam_size) {
+    std::vector<std::pair<double, Factor>> empty;
+    for (const Factor& factor : query_->factors_) {
+      std::vector<int> inf_vars;
+      Factor giv_labels;
+      for (auto var = factor.begin(); var != factor.end(); var++) {
+        Assignment a= assignments_[(*var)];
+        if (a.must_infer == true) {
+          inf_vars.push_back((*var));
+        } else {
+          giv_labels.insert(a.label);
+        }
+      }
+      if (inf_vars.size() <= 1) {
+        continue;
+      }
+
+      std::vector<Factor> factors;
+      GetFactorCandidates(fweights, factor.size(), &factors, giv_labels, beam_size);
+      double best_score = 0;
+      std::vector<int> best_assignments;
+      for (auto var = inf_vars.begin(); var != inf_vars.end(); var++) {
+        best_score += GetNodeScore(fweights, *var);
+        best_assignments.push_back(assignments_[*var].label);
+      }
+      std::vector<Factor> factors_candidates;
+      for (uint j = 0; j < factors.size(); j++) {
+        int factor_matches_giv_vars = true;
+        for (auto label = giv_labels.begin(); label != giv_labels.end(); label++) {
+          if (factors[j].count(*label) == 0) {
+            factor_matches_giv_vars = false;
+            break;
+          }
+        }
+        if (factor_matches_giv_vars) {
+          factors_candidates.push_back(factors[j]);
+        }
+      }
+      for (uint j = 0; j < factors_candidates.size(); j++) {
+        Factor current_candidate = factors_candidates[j];
+        std::set<int> distinct_labels;
+        std::vector<int> candidate_inf_labels;
+        std::unordered_set<std::vector<int>> permutations;
+        for (auto label = current_candidate.begin(); label != current_candidate.end(); label++) {
+          if (giv_labels.count(*label) == 0) {
+            candidate_inf_labels.push_back(*label);
+            distinct_labels.insert(*label);
+          }
+        }
+        uint64 num_permutations = CalculateFactorial(distinct_labels.size());
```

Comment:
Yeah, I will put a limit to stop not going randomly forever. The overflow is handled in the following if saying that if the distinct_label.size() is equal to 64 then the integer has overflown so I go directly with random permutation.